File: python/data-structures/point2d.py
import math

# point(number, number) -> tuple
# Constructs a two dimensional point as a tuple.
# point(1, 2) -> (1, 2)

def point(x, y):
    return (x, y)

# distance(tuple, tuple) -> float
# Calculates the distance between two points.
# distance((3, 0), (2, 0)) -> 1.0

def distance(p1, p2):
    x1, y1 = p1 
    x2, y2 = p2
    return math.hypot(x2 - x1, y2 - y1)

# path(tuple) -> tuple
# Constructs a path as a tuple of points.
# path(point(1, 2), point(3, 4)) -> ((1, 2), (3, 4))

def path(*points):
    return points

# path_length(tuple) -> float
# Computes the length of a path along a series of points.
# path_length(path(point(1, 1), point(5, 1), point(5, 4), point(1, 1))) -> 12.0

def path_length(path):
    sum = 0
    length = len(path)
    for i in range(length - 1):
        p1 = path[i]
        p2 = path[i + 1]
        sum += distance(p1, p2)
    return sum

File: geometry/golang/points/point2d.go
package points

import (
	"fmt"
	"math"
)

// The maximum allowable difference in precision between two floating-point coordinates.
const epsilon = 0.000001

type Point2 struct {
	X, Y float64
}

// Fulfills the Stringer interface for the fmt package.
func (p Point2) String() string {
	x := p.X
	y := p.Y
	return fmt.Sprintf("pt(%.2f, %.2f)", x, y)
}

// Calculates the distance between two points.
func (p1 Point2) Segment(p2 Point2) float64 {
	x1 := p1.X
	y1 := p1.Y
	x2 := p2.X
	y2 := p2.Y
	return math.Hypot(x2-x1, y2-y1)
}

// A geometric path implemented as a series of points.
type Path []Point2

func (p Path) Length() float64 {
	sum := float64(0)
	for i := 0; i < (len(p) - 1); i += 1 {
		p1 := p[i]
		p2 := p[i+1]
		sum += p1.Segment(p2)
	}
	return sum
}

// Outputs a point with the rounded components of the receiver point.
func (p Point2) Round() Point2 {
	x := math.Round(p.X)
	y := math.Round(p.Y)
	return Point2{x, y}
}
